core/trino-main/src/main/java/io/trino/spiller/FileSingleStreamSpiller.java

```diff
@@ -137,61 +177,176 @@ public long getSpilledPagesInMemorySize()
+                                fileIndex = (fileIndex + 1) % fileCount;
+                                found = true;
+                                break;
+                            }
+                            fileIndex = (fileIndex + 1) % fileCount;
+                        } while (fileIndex != start);
+
+                        if (!found) {
+                            return endOfData();
+                        }
+                    }
+                }
+            };
+        }
+        catch (IOException e) {
+            fileSystemErrorHandler.run();
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, "Failed to read spilled pages", e);
+        }
     }
 
     @Override
     public ListenableFuture<List<Page>> getAllSpilledPages()
     {
-        return executor.submit(() -> ImmutableList.copyOf(getSpilledPages()));
+        checkNoSpillInProgress();
+        checkState(writable.getAndSet(false), "Repeated reads are disallowed to prevent potential resource leaks");
+
+        Optional<SecretKey> encryptionKey = this.encryptionKey;
+        checkState(encrypted == encryptionKey.isPresent(), "encryptionKey has been discarded");
+        this.encryptionKey = Optional.empty();
+
+        List<ListenableFuture<List<List<Page>>>> futures = new ArrayList<>();
+        for (FileHolder file : targetFiles) {
+            futures.add(executor.submit(() -> {
+                PageDeserializer deserializer = serdeFactory.createDeserializer(encryptionKey);
+                ImmutableList.Builder<List<Page>> segments = ImmutableList.builder();
+                try (Closer closer = Closer.create()) {
+                    Iterator<Iterator<Page>> segmentIterator = readSegments(deserializer, file, closer);
+                    // materialize all segments from the file using executor thread
+                    while (segmentIterator.hasNext()) {
+                        segments.add(ImmutableList.copyOf(segmentIterator.next()));
+                    }
+                }
+                return segments.build();
+            }));
+        }
+
+        // Combine pages from all spill files according to the segment order.
+        return Futures.transform(Futures.allAsList(futures), segmentsPerFile -> {
+            ImmutableList.Builder<Page> builder = ImmutableList.builder();
+            int fileCount = segmentsPerFile.size();
+            List<Iterator<List<Page>>> iterators = new ArrayList<>(fileCount);
+            for (List<List<Page>> segments : segmentsPerFile) {
+                iterators.add(segments.iterator());
+            }
+
+            int fileIndex = 0;
+            while (true) {
+                boolean found = false;
+                for (int i = 0; i < fileCount; i++) {
+                    int index = (fileIndex + i) % fileCount;
+                    Iterator<List<Page>> segments = iterators.get(index);
+                    if (segments.hasNext()) {
+                        segments.next().forEach(builder::add);
```

Comment:
Probably: `builder.addAll(segments.next())` to allow bulk expansion

Reply:
That is a good idea